pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,255 @@
+		}
+
+		err = m.dbusCon.ReloadLogindConf()
+		if err != nil {
+			return err
+		}
+
+		// Read the current inhibitDelay again, if the override was successful, currentInhibitDelay will be equal to shutdownGracePeriodRequested.
+		updatedInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+		if err != nil {
+			return err
+		}
+
+		if updatedInhibitDelay != m.shutdownGracePeriodRequested {
+			return fmt.Errorf("node shutdown manager was unable to update logind InhibitDelayMaxSec to %v (ShutdownGracePeriod), current value of InhibitDelayMaxSec (%v) is less than requested ShutdownGracePeriod", m.shutdownGracePeriodRequested, updatedInhibitDelay)
+		}
+	}
+
+	err = m.aquireInhibitLock()
+	if err != nil {
+		return err
+	}
+
+	events, err := m.dbusCon.MonitorShutdown()
+	if err != nil {
+		releaseErr := m.dbusCon.ReleaseInhibitLock(m.inhibitLock)
+		if releaseErr != nil {
+			return fmt.Errorf("failed releasing inhibitLock: %v and failed monitoring shutdown: %v", releaseErr, err)
+		}
+		return fmt.Errorf("failed to monitor shutdown: %v", err)
+	}
+
+	go func() {
+		// Monitor for shutdown events. This follows the logind Inhibit Delay pattern described on https://www.freedesktop.org/wiki/Software/systemd/inhibit/
+		// 1. When shutdown manager starts, an inhibit lock is taken.
+		// 2. When shutdown(true) event is received, process the shutdown and release the inhibit lock.
+		// 3. When shutdown(false) event is received, this indicates a previous shutdown was cancelled. In this case, acquire the inhibit lock again.
+		for {
+			select {
+			case isShuttingDown := <-events:
+				klog.V(1).Infof("Shutdown manager detected new shutdown event, isNodeShuttingDownNow: %t", isShuttingDown)
+
+				m.nodeShuttingDownMutex.Lock()
+				m.nodeShuttingDownNow = isShuttingDown
+				m.nodeShuttingDownMutex.Unlock()
+
+				if isShuttingDown {
+					m.processShutdownEvent()
+				} else {
+					m.aquireInhibitLock()
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *Manager) aquireInhibitLock() error {
+	lock, err := m.dbusCon.InhibitShutdown()
+	if err != nil {
+		return err
+	}
+	m.inhibitLock = lock
+	return nil
+}
+
+// ShutdownStatus will return an error if the node is currently shutting down.
+func (m *Manager) ShutdownStatus() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+
+	m.nodeShuttingDownMutex.Lock()
+	defer m.nodeShuttingDownMutex.Unlock()
+
+	if m.nodeShuttingDownNow {
+		return fmt.Errorf("node is shutting down")
+	}
+	return nil
+}
+
+func (m *Manager) processShutdownEvent() error {
+	klog.V(1).Infof("Shutdown manager processing shutdown event")
+	activePods := m.getPods()
+
+	nonCriticalPodGracePeriod := m.shutdownGracePeriodRequested - m.shutdownGracePeriodCriticalPods
+
+	var wg sync.WaitGroup
+	wg.Add(len(activePods))
+	for _, pod := range activePods {
+		go func(pod *v1.Pod) {
+			defer wg.Done()
+
+			var gracePeriodOverride int64
+			if kubelettypes.IsCriticalPod(pod) {
```

Comment:
I commented in the KEP https://github.com/kubernetes/enhancements/pull/2001/files#r565507971 but I don't think this is appropriate.  It forces an administrator deploying a system infrastructure pod to use one of these two priority classes to get this behavior, which means you can't create new priority classes for your critical infrastructure pods to control the order of eviction.  

Also, I expect all static pods to be covered by this logic, because it is impossible to drain a static pod correctly outside the kubelet (since kubelet controls the lifecycle, not an outside entity).

Reply:
kube-apiserver depends on SDN (for aggregation and for webhooks). If both are implemented as static pods, and SDN is able to go through a LB to reach an apiserver on another node, how can we make sure SDN stays up longer than the static kube-apiserver pod?

Reply:
see https://github.com/kubernetes/enhancements/pull/2001#issuecomment-770072076 for discussion